### PR TITLE
feat: update description with append model

### PR DESCRIPTION
**Pull request by ReedSoftware**

### CODE REVIEW

File: server/routes/journeys/referral/description/schemas.ts [hunk 1]
```diff
@@ -1,18 +1,22 @@
...
 import { Request } from 'express'
 import { createSchema } from '../../../../middleware/validationMiddleware'
+import { getMaxCharsAndThreshold } from '../../../../utils/appendFieldUtils'
 
-const DESCRIPTION_OF_CONCERN_TOO_LONG_MSG = 'Description must be 4,000 characters or less'
+const DESCRIPTION_OF_CONCERN_TOO_LONG_MSG = (length: number) =>
+  `Description must be ${length.toLocaleString()} characters or less`
 
 export const schemaFactory = async (req: Request) => {
+  const { maxLengthChars } = getMaxCharsAndThreshold(req, req.journeyData.csipRecord!.referral.descriptionOfConcern)
```
Comment (bill-moj) [inline server/routes/journeys/referral/description/schemas.ts]: `[Node] TypeError: Cannot read properties of undefined (reading 'referral')` in MaR journey

File: server/routes/journeys/referral/description/view.njk [hunk 2]
```diff
@@ -35,18 +36,23 @@
...
...aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
```
Comment (bill-moj) [inline server/routes/journeys/referral/description/view.njk:58]: aaaaaaaaaaaaaaaaa!

File: server/routes/journeys/update-referral/description/controller.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  }
+
+  checkSubmitToAPI = async (req: Request<unknown, unknown, SchemaType>, res: Response, next: NextFunction) =>
+    this.submitChanges({
+      req,
+      res,
+      next,
+      changes: {
+        descriptionOfConcern: generateSaveTimestamp(req.journeyData.prisoner!) + req.body.descriptionOfConcern,
+      },
```
Comment (bill-moj) [inline server/routes/journeys/update-referral/description/controller.ts:25]: backend api doesn't append descriptionOfConcern. we have to append/concatenate the content here. 
so it's going to be `req.journeyData.referral!.descriptionOfConcern + generateSaveTimestamp(req.journeyData.prisoner!) + req.body.descriptionOfConcern,`

**Review by bill-moj — CHANGES_REQUESTED** [review]

**Review by bill-moj — DISMISSED** [review]

File: server/utils/appendFieldUtils.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+/**
+ * If the description + divider test is equal to or more than 3000 characters, then show the character count
+ * If the description + divider text is less than 3000 characters, do not show the character count until the user reaches a total of 3000 characters for all entries plus dividers
+ */
+export const generateSaveTimestamp = (prisoner: Pick<Prisoner, 'firstName' | 'lastName'>) =>
+  `\n\n[On ${formatDisplayDateTime(new Date().toString())}, ${lastNameCommaFirstName(prisoner)} added:]\n\n`
+
+export const getMaxCharsAndThreshold = (req: Request, appendField?: string) => {
+  const timestampLength = generateSaveTimestamp(req.journeyData.prisoner!).length
```
Comment (iterative-it) [inline server/utils/appendFieldUtils.ts:14]: This should be the display name of the logged in user, not the prisoner's name. [On time and date, userLastname, userFirstname added:]

File: server/utils/appendFieldUtils.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import { formatDisplayDateTime } from './datetimeUtils'
+import { lastNameCommaFirstName } from './miniProfileUtils'
+import Prisoner from '../services/prisonerSearch/prisoner'
+
+/**
+ * If the description + divider test is equal to or more than 3000 characters, then show the character count
+ * If the description + divider text is less than 3000 characters, do not show the character count until the user reaches a total of 3000 characters for all entries plus dividers
+ */
+export const generateSaveTimestamp = (prisoner: Pick<Prisoner, 'firstName' | 'lastName'>) =>
+  `\n\n[On ${formatDisplayDateTime(new Date().toString())}, ${lastNameCommaFirstName(prisoner)} added:]\n\n`
```
Comment (iterative-it) [inline server/utils/appendFieldUtils.ts:11]: Design has this as "lastname firstname" not "firstname, lastname" unless that's changed post ticketing.

Comment (ReedSoftware) [reply server/utils/appendFieldUtils.ts:11]: image

Comment (ReedSoftware) [reply server/utils/appendFieldUtils.ts:11]: I can clarify with Matt if this is clashing with the designs though

Comment (ReedSoftware) [reply server/utils/appendFieldUtils.ts:11]: I assume this should be `res.locals.user.displayName.substring(0, 240)` as per `area-of-work` assuming designs are correct?

Comment (iterative-it) [reply server/utils/appendFieldUtils.ts:11]: Just `res.locals.user.displayName`. No need to truncate it as the other length checks will account for it.

**Review by bill-moj — APPROVED** [review]

**Pull request merged**